apps/champions/lib/champions/battle/simulator.ex

```diff
@@ -63,187 +64,459 @@ defmodule Champions.Battle.Simulator do
     team_2 = Enum.into(team_2, %{}, fn unit -> create_unit_map(unit, 2) end)
     units = Map.merge(team_1, team_2)
 
+    initial_state = %{units: units, skills_being_cast: [], pending_effects: [], pending_executions: []}
+
     # The initial_step_state is what allows the battle to be simultaneous. If we refreshed the accum on every action,
     # we would be left with a turn-based battle. Instead we take decisions based on the state of the battle at the beggining
     # of the step regardless of the changes that happened "before" (execution-wise) in this step.
-    Enum.reduce_while(1..@maximum_steps, units, fn step, initial_step_state ->
+    Enum.reduce_while(1..@maximum_steps, initial_state, fn step, initial_step_state ->
+      initial_step_state = Map.put(initial_step_state, :step_number, step)
+
       new_state =
-        initial_step_state
-        |> Enum.map(fn {id, _unit} -> id end)
-        |> Enum.reduce(initial_step_state, fn unit_id, current_state ->
-          process_step(initial_step_state[unit_id], initial_step_state, current_state)
+        Enum.reduce(initial_step_state.units, initial_step_state, fn {unit_id, unit}, current_state ->
+          Logger.info("Process step #{step} for unit #{format_unit_name(unit)}")
+          process_step_for_unit(initial_step_state.units[unit_id], current_state, initial_step_state)
         end)
 
+      new_state =
+        Enum.reduce(new_state.skills_being_cast, new_state, fn skill, current_state ->
+          Logger.info("Process step #{step} for skill #{skill.name} cast by #{String.slice(skill.caster_id, 0..2)}")
+
+          # We need the initial_step_state to decide effect targets
+          process_step_for_skill(skill, current_state, initial_step_state)
+        end)
+
+      new_state =
```

Comment:
Since we're updating the same map multiple times i think a best approach would be create a bigger function for each step that traverse the players/skills/effects and do what it needs to do (you can check how `apps/arena/lib/arena/game_updater.ex` handles the game update in line 103)

Reply:
Done in 0af43030! It adds more wrapper functions though 🤔

Reply:
yeah, i think the change is worth tho. 

```
initial_step_state
        |> Map.put(:step_number, step)
        |> process_step_for_units()
        |> process_step_for_skills(initial_step_state)
        |> process_step_for_effects()
```

it's a lot more readable than the previous iteration imo! 🪨